src/librustc_lint/types.rs

```diff
@@ -182,37 +223,77 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for TypeLimits {
+                                            parent_expr.span,
+                                            "only u8 can be casted into char",
+                                        );
+                                        err.span_suggestion(
+                                            parent_expr.span,
+                                            &"use a char literal instead",
+                                            format!("'\\u{{{:X}}}'", lit_val),
+                                        );
                                         err.emit();
-                                        return
+                                        return;
                                     }
                                 }
                             }
-                            cx.span_lint(OVERFLOWING_LITERALS,
-                                         e.span,
-                                         &format!("literal out of range for {:?}", t));
+                            if let Some(repr_str) = get_bin_hex_repr(cx, lit) {
+                                let bits = uint_ty_bits(t, cx.sess().target.usize_ty);
+                                // u128 cannot be greater than max -> compiler error
+                                let actually = lit_val & ((1 << bits) - 1);
+                                let mut err = cx.struct_span_lint(
+                                    OVERFLOWING_LITERALS,
+                                    e.span,
+                                    &format!("literal out of range for {:?}", t),
+                                );
+                                err.note(&format!(
+                                    "the literal `{}` (decimal `{}`) does not fit into \
+                                     an `{:?}` and will become `{}{:?}`.",
+                                    repr_str, lit_val, t, actually, t
+                                ));
+                                let sugg_ty = get_fitting_type(
+                                    &cx.tables.node_id_to_type(e.hir_id).sty,
+                                    lit_val,
+                                    false,
+                                ).map_or(
+                                    String::new(),
+                                    |ty| {
+                                        if let ty::TyUint(t) = ty {
+                                            format!("Consider using `{:?}`", t)
+                                        } else {
+                                            String::new()
+                                        }
+                                    },
+                                );
+                                if !sugg_ty.is_empty() {
+                                    err.help(&sugg_ty);
+                                }
+
+                                err.emit();
+                                return;
+                            }
+                            cx.span_lint(
```

Comment:
It feels like there's a lot of duplication between the error reporting of signed and unsigned integers. Is it possible to merge them into a function?

Reply:
I think changing the function get_fitting_type(), that it returns the suggestion string directly is the best approach here. 
An extra function for the `err.note()` part is hard, because we need the `actually` value as `i128` in the singed case and as `u128` in the unsigned case. Same problem with the Type.